deeppavlov/models/vectorizers/hashing_tfidf_vectorizer.py

```diff
@@ -0,0 +1,173 @@
+"""
+Copyright 2017 Neural Networks and Deep Learning lab, MIPT
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from collections import Counter
+from typing import List, Any, Generator, Tuple, KeysView, ValuesView, Type, Dict
+
+import scipy as sp
+from scipy import sparse
+import numpy as np
+from sklearn.utils import murmurhash3_32
+
+from deeppavlov.models.tokenizers.spacy_tokenizer import StreamSpacyTokenizer
+from deeppavlov.core.models.estimator import Estimator
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.core.common.registry import register
+
+TOKENIZER = None
+logger = get_logger(__name__)
+
+
+def hash_(token, hash_size):
+    return murmurhash3_32(token, positive=True) % hash_size
+
+
+@register('hashing_tfidf_vectorizer')
+class HashingTfIdfVectorizer(Estimator):
+    """
+    Create a tfidf matrix from collection of documents.
+    """
+
+    def __init__(self, hash_size=2 ** 24, tokenizer: Type = StreamSpacyTokenizer, doc_index: dict =None,
+                 save_path: str = None, load_path: str = None, **kwargs):
+        """
+
+        :param hash_size: a size of hash, power of 2
+        :param tokenizer: a tokenizer class
+        """
+        super().__init__(save_path=save_path, load_path=load_path)
+
+        self.hash_size = hash_size
+        self.tokenizer = tokenizer
+        self.term_freqs = None
+        self.doc_index = doc_index
+
+        global TOKENIZER
+        TOKENIZER = self.tokenizer
+
+        self.rows = []
+        self.cols = []
+        self.data = []
+
+    def __call__(self, question: str) -> sp.sparse.csr_matrix:
```

Comment:
Components are expected to work with batches, not individual instances

Reply:
Done